src/Features/LanguageServer/Protocol/Workspaces/LspWorkspaceManager.cs

```diff
@@ -23,46 +25,32 @@ namespace Microsoft.CodeAnalysis.LanguageServer;
+/// 
+/// Doing the forking like this has a few nice properties.
 /// <list type="bullet">
-///   <item>LSP didChange events only cause us to update the document that changed, not all open documents.</item>
-///   <item>Since we incrementally update our LSP documents, we only have to re-parse the document that changed.</item>
-///   <item>Since we incrementally update our LSP documents, project versions for other open documents remain unchanged.</item>
-///   <item>We are not reliant on the workspace being updated frequently (which it is not in VSCode) to do checksum diffing between LSP and the workspace.</item>
+///   <item>99% of the time the VS workspace matches the LSP text.  In those cases we do 0 re-parsing, share compilations, versions, checksum calcs, etc.</item>
+///   <item>In the 1% of the time that we do not match, we can simply and easily compute a fork.</item>
+///   <item>The code is relatively straightforward</item>
 /// </list>
 /// </remarks>
 internal class LspWorkspaceManager : IDocumentChangeTracker, IDisposable
 {
     /// <summary>
-    /// Indicates whether the LSP solution has changed in a non-tracked document context.
+    /// A cache from workspace to the last solution we returned for LSP.
+    /// When this solution comes from a fork operation we store the workspace version it was forked from.
     /// 
-    /// <b>IMPORTANT:</b> Implementations of this event handler should do as little synchronous work as possible since this will block.
+    /// Access to this is gauranteed to be serial by the <see cref="RequestExecutionQueue"/>
     /// </summary>
-    public EventHandler<WorkspaceChangeEventArgs>? LspSolutionChanged;
-
-    /// <summary>
-    /// Lock to gate access to the <see cref="_workspaceToLspSolution"/> and <see cref="_trackedDocuments"/>
-    /// Access from the LSP server is serial as the LSP queue is processed serially until
-    /// after we give the solution to the request handlers.  However workspace events can interleave
-    /// so we must protect against concurrent access here.
-    /// </summary>
-    private readonly object _gate = new();
-
-    /// <summary>
-    /// A map from the registered workspace to the running lsp solution with incremental changes from LSP
-    /// text sync applied to it.
-    /// All workspaces registered by the <see cref="LspWorkspaceRegistrationService"/> are added as keys to this dictionary
-    /// with a null value (solution) initially.  As LSP text changes come in, we create the incremental solution from the workspace and update it with changes.
-    /// When we detect a change that requires us to re-fork the LSP solution from the workspace we null out the solution for the key.
-    /// </summary>
-    private readonly Dictionary<Workspace, Solution?> _workspaceToLspSolution = new();
+    private readonly Dictionary<Workspace, (int? ForkedFromVersion, Solution Solution)> _cachedLspSolutions = new();
```

Comment:
1. when is ForkedFromVersion null?
2. nit: tuple parameters are normally camelCased in IDE.